doc/swish/parallel.tex

```diff
@@ -0,0 +1,183 @@
+order. Control of start order provides a mechanism to help detect and
+test for assumptions and errors in user-code.}. As each worker
+completes, a new one starts. When all the workers complete, the kernel
+returns an API-specific value.
+
+The kernel also supports a \code{timeout} for the entire parallel
+operation. When the timeout expires all remaining workers are killed
+with the reason \code{timeout} and an exception is thrown in the
+caller process.
+
+Because the caller is not killed by the kernel, it acts as an observer
+of the kernel's behavior. As such, when an \code{event-mgr} process
+exists, the caller publishes \code{<child-start>} and
+\code{<child-end>} events described in
+Section~\ref{supervisor:events}. Because the potential number of
+workers may be high, the worker processes are not logged.
+
+\section {Programming Interface}
+
+Kernel options can be defined using
+\code{(parallel:options [\var{option} \var{value}] \etc)}.
+The following options may be used:
+\defineentry{parallel:options}
+\phantomsection % make pageref go to correct page for this label
+\label{parallel:options}
+
+\begin{tabular}{lp{5em}p{.65\textwidth}}
+  option & default & description \\ \hline
+
+  \code{start-limit}
+  & \code{(most-positive-fixnum)}
+  & a positive fixnum; the number of workers allowed to start and run
+  concurrently \\
+
+  \code{order}
+  & \code{random}
+  & a symbol, one of \code{random}, \code{left}, or \code{right} \\
+
+  \code{timeout}
+  & \code{infinity}
+  & \code{infinity} or a nonnegative fixnum; the number of
+  milliseconds the workers are allowed to execute before failing the
+  entire operation \\
+\end{tabular}
+
+\defineentry{parallel}
+\begin{syntax}
+  \code{(parallel () \var{e} \etc)}\\
+  \code{(parallel ([\var{option} \var{value}] \etc) \var{e} \etc)}\\
+  \code{(parallel ,\var{options} \var{e} \etc)}
+\end{syntax}
+
+The \code{parallel} construct invokes each expression \var{e}, \etc{}
+concurrently and returns a list of the resulting values. Expressions
+may run concurrently unless \code{start-limit} is 1.
+
+The \code{(parallel ([\var{option} \var{value}] \etc) \var{e} \etc)}
+form constructs a copy of the default options overriding the specified
+\var{option} with the specified \var{value}.
+
+\defineentry{parallel"!}
+\begin{syntax}
+  \code{(parallel! () \var{e} \etc)}\\
+  \code{(parallel! ([\var{option} \var{value}] \etc) \var{e} \etc)}\\
+  \code{(parallel! ,\var{options} \var{e} \etc)}
+\end{syntax}
+
+The \code{parallel!} construct behaves as \code{parallel} but returns
+an unspecified value.
+
+\defineentry{parallel:execute}
+\begin{procedure}
+  \code{(parallel:execute \opt{\var{options}} \var{thunks})}
+\end{procedure}
+\returns{} a list of results
+
+The \code{parallel:execute} procedure takes a list of \var{thunks}
+(procedures of no arguments). \code{parallel:execute} returns a list
+of resulting values. Each \var{procedure} may run concurrently unless
+\code{start-limit} is 1.
+
```

Comment:
Describe the optional "options" argument.

Reply:
Updating as suggested.